### DISCUSSION TITLE
My extension and/or script isn't being properly loaded

### DISCUSSION THREAD

**Discussion by dayunbao** [Mikubill/sd-webui-controlnet#2117] [Q&A]
I'm moving this here from the webui repo's Q&A since no one has answered, and I am basing my extension on this one:

This is my first time creating an extension, and I'm having some problems.  What I'd like to do is expose some of my own endpoints like is done here https://github.com/Mikubill/sd-webui-controlnet/wiki/API.  For example, `GET /controlnet/model_list`.  So I based my extension on that one, and have this directory structure:

```
my_extension
|_scripts
   |_api.py
   |_my_script.py
```
Here's what `api.py` looks like:
```
from fastapi import FastAPI
from modules import script_callbacks

def my_extension_api(_, app: FastAPI):
    @app.get('my_extension/version')
    def version():
        return {'version': 1}


script_callbacks.on_app_started(my_extension_api)
```
and `my_script.py`:
```
from modules import scripts

class Script(scripts.Script):
    def __init__(self):
        super().__init__()

    def title(self):
        return "My Extension"

    def show(self, is_img2img):
        return scripts.AlwaysVisible
```
I've tried dropping the extension directly into the `extensions` folder and installing from URL.  The extension is marked as active.  I also do not gt any errors in the logs when starting the server.  However, I get a 404 when trying to reach the endpoint, and if I call `GET /sdapi/v1/scripts`, my script isn't shown.  Any suggestions?

**Comment by ljleb** [comment]
First of all, please do not put more than one python file under the scripts folder. The webui imports all files individually on top of your local imports, so it can create confusing situation. Ideally you have a single file under scripts, and you put everything else inside a separate directory, i.e. `webui/extensions/your-extension/lib_your_extension/...` from which you can import.

The single file under scripts/ will be executed when the webui loads your extension, so any other file or module this file imports will also get executed. Using this mechanism, you can then import your api from your custom dir if you dont want to put everything in the same file.

I'm not sure yet why it does not seem to work for you otherwise. Did you add the cli flag `--api`? the webui needs this to be able to extend the api. If you put a breakpoint in the callback, is it ever reached?

**Reply by dayunbao** [reply]
I made a really stupid mistake in my code, and that was causing the problems.

Thank for you help!
